### PR TITLE
Add Telit ME310G1 LTE modem, add shutdown script to modem_cellular

### PR DESCRIPTION
**Pull request by Erlkoenig90**
Add support for the Telit ME310G1 LTE modem which is very similar to the ME910G1 but lacks the reset line (binding adjusted accordingly). Re-use the ME910G1's AT command scripts.

Also add support to define a shutdown script, i.e. AT commands to shutdown the modem and define one for the ME310G1 (`AT#SHDN`). This allows us to shutdown the modem much quicker compared to using a power pulse which saves power. This probably also works for the ME910G1 but since I can't test it I didn't add it.

### CODE REVIEW

**Review by tomi-font — CHANGES_REQUESTED** [review]
Please split your commit into two: one for the addition of the shutdown script, and another for the addition of the ME310G1 modem. You'll also need to change the formatting of the commit, and it would be nice that you include some description text to the commit message as you wrote in the PR description.
I have triggered the CI jobs that needed approval so you should see compliance failures that you'll need to address.

File: drivers/modem/modem_cellular.c [hunk 1]
```diff
@@ -1089,7 +1096,6 @@ static int modem_cellular_on_carrier_on_state_leave(struct modem_cellular_data *
 
 static int modem_cellular_on_init_power_off_state_enter(struct modem_cellular_data *data)
 {
-	modem_pipe_close_async(data->uart_pipe);
 	modem_cellular_start_timer(data, K_MSEC(2000));
```
[the remark this answers is not in the document]
Comment (Erlkoenig90) [reply drivers/modem/modem_cellular.c:1099]: According to the discussion on discord, it is necessary to allow PPP to shutdown.

File: drivers/modem/modem_cellular.c [hunk 2]
```diff
@@ -2098,7 +2174,7 @@ MODEM_CHAT_SCRIPT_DEFINE(swir_hl7800_dial_chat_script, swir_hl7800_dial_chat_scr
 			 dial_abort_matches, modem_cellular_chat_callback_handler, 10);
 #endif
 
-#if DT_HAS_COMPAT_STATUS_OKAY(telit_me910g1)
+#if DT_HAS_COMPAT_STATUS_OKAY(telit_me910g1) || DT_HAS_COMPAT_STATUS_OKAY(telit_me310g1)
```
[the remark this answers is not in the document]
Comment (tomi-font) [reply drivers/modem/modem_cellular.c:2175]: There is still this that I think would be good to address.

**Review by tomi-font — CHANGES_REQUESTED** [review]
I think that you'll want to at least add `modem_cellular:` to your commits' titles. You can find out what to use by checking what people have done before: `git log drivers/modem/modem_cellular.c`

Also, you still have some compliance checks failing: https://github.com/zephyrproject-rtos/zephyr/actions/runs/11106324597/job/30900872718?pr=78890

File: drivers/modem/modem_cellular.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (Erlkoenig90) [reply drivers/modem/modem_cellular.c:2175]: Do you mean I should rename the scripts to something like `telit_mex10g1_init_chat_script`?

**Comment by Erlkoenig90** [comment]
> Also, you still have some compliance checks failing: https://github.com/zephyrproject-rtos/zephyr/actions/runs/11106324597/job/30900872718?pr=78890

I can only fix all those issues, by reformatting the *entire* file (as the compliance check apparently complains of about *unchanged* lines adjacent to changed ones). Should I really do that?

File: drivers/modem/modem_cellular.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (tomi-font) [reply drivers/modem/modem_cellular.c:2175]: Yeah, this naming would be fine to me. Really just something to make it clear that it's not just for the 910.

**Comment by tomi-font** [comment]
> I can only fix all those issues, by reformatting the _entire_ file (as the compliance check apparently complains of about _unchanged_ lines adjacent to changed ones). Should I really do that?

I'd suggest you fix the issues outlined [here](https://github.com/zephyrproject-rtos/zephyr/actions/runs/11106324597/job/30900872718?pr=78890). There aren't too many of them, and hopefully they won't grow in number. Or maybe you could just use `clang-format` to directly format the whole file to the new required format (as a separate commit).
I think the only alternative to this would be to have the PR merged with a failing CI, which I don't know if it would be accepted. Would need to ask in Discord.
I agree it's unfortunate that this comes up in PRs that don't modify those lines. cc @kartben

**Review by tomi-font — COMMENTED** [review]
The commits now look good, but seems like you didn't push any changes to `drivers/modem/modem_cellular.c`?

**Comment by Erlkoenig90** [comment]
> The commits now look good, but seems like you didn't push any changes to `drivers/modem/modem_cellular.c`?

Yes, I'm going to ask for help about the formatting in #pr-help when the compliance check is run. Particularly the [last change](https://github.com/zephyrproject-rtos/zephyr/actions/runs/11106324597/job/30900872718?pr=78890) suggested by the last compliance check doesn't make sense at all - it wants me to delete some random lines of code? Or does it mean to indicate I deleted those lines, which I didn't do?

**Comment by tomi-font** [comment]
> Yes, I'm going to ask for help about the formatting in #pr-help when the compliance check is run.

Okay. Please also don't forget to address or respond to [my last remaining open comment](https://github.com/zephyrproject-rtos/zephyr/pull/78890/files#r1795186523).

> Particularly the [last change](https://github.com/zephyrproject-rtos/zephyr/actions/runs/11106324597/job/30900872718?pr=78890) suggested by the last compliance check doesn't make sense at all - it wants me to delete some random lines of code? Or does it mean to indicate I deleted those lines, which I didn't do?

Yeah, weird indeed. It proposes diffs, so lines starting with `-` are to be deleted and replaced with those starting with `+`.
It seems that the last one is missing the `+` part. Maybe the output got somehow truncated at 100 lines? I would expect that it wants the same formatting changes on `MODEM_CELLULAR_DEFINE_AND_INIT_USER_PIPES()` than for the second to last change.

File: drivers/modem/modem_cellular.c [hunk 3]
```diff
@@ -2098,8 +2173,8 @@ MODEM_CHAT_SCRIPT_DEFINE(swir_hl7800_dial_chat_script, swir_hl7800_dial_chat_scr
 			 dial_abort_matches, modem_cellular_chat_callback_handler, 10);
 #endif
 
-#if DT_HAS_COMPAT_STATUS_OKAY(telit_me910g1)
-MODEM_CHAT_SCRIPT_CMDS_DEFINE(telit_me910g1_init_chat_script_cmds,
+#if DT_HAS_COMPAT_STATUS_OKAY(telit_me910g1) || DT_HAS_COMPAT_STATUS_OKAY(telit_me310g1)
+MODEM_CHAT_SCRIPT_CMDS_DEFINE(telit_mex10g1_init_chat_script_cmds,
 				  MODEM_CHAT_SCRIPT_CMD_RESP_NONE("AT", 100),
 				  MODEM_CHAT_SCRIPT_CMD_RESP_NONE("AT", 100),
 				  MODEM_CHAT_SCRIPT_CMD_RESP_NONE("AT", 100),
```
Comment (tomi-font) [inline drivers/modem/modem_cellular.c:2180]: There is still a comment about the `me910g1` some lines below. Do you know if this affects the 310 as well? (We can keep this just as it is.)

**Review by tomi-font — APPROVED** [review]
Great!

File: drivers/modem/modem_cellular.c (hunk 3 above)
Comment (Erlkoenig90) [reply drivers/modem/modem_cellular.c:2180]: I didn't test it differently, it just worked as is...

**Review by bjarki-andreasen — APPROVED** [review]

**Review by rerickson1 — CHANGES_REQUESTED** [review]
Sorry for this nit, but I think the DTS binding addition should be its own commit.

**Comment by Erlkoenig90** [comment]
I renamed the modem scripts as requested. Even though the compliance check [reports an error](https://github.com/zephyrproject-rtos/zephyr/actions/runs/11363320129/job/31607287771?pr=78890#step:9:697), it counts as successful, even though exactly the same formatting was reported as an error in the last tries? [According](https://discord.com/channels/720317445772017664/1296044091921928273/1296045105509040128) to @pdgendt we can ignore the formatting issues.

**Comment by Erlkoenig90** [comment]
> Sorry for this nit, but I think the DTS binding addition should be its own commit.

@rerickson1 What would be the correct order of commits?

1. Add support for shutdown script
2. Add ME310 C code
3. Add ME310 DTS binding

Or swap 2 and 3?

**Comment by rerickson1** [comment]
> > Sorry for this nit, but I think the DTS binding addition should be its own commit.
> 
> @rerickson1 What would be the correct order of commits?
> 
> 1. Add support for shutdown script
> 2. Add ME310 C code
> 3. Add ME310 DTS binding
> 
> Or swap 2 and 3?

You need the DTS binding to use it in the driver, so it should be first in time.

**Comment by Erlkoenig90** [comment]
> > > Sorry for this nit, but I think the DTS binding addition should be its own commit.
> > 
> > 
> > @rerickson1 What would be the correct order of commits?
> > 
> > 1. Add support for shutdown script
> > 2. Add ME310 C code
> > 3. Add ME310 DTS binding
> > 
> > Or swap 2 and 3?
> 
> You need the DTS binding to use it in the driver, so it should be first in time.

The commit is now split in 2.

**Comment by rerickson1** [comment]
Change commit title from `modem_cellular: Add DTS binding for Telit ME310G1` to `dts: bindings: Add binding for Telit ME310G1`

**Review by rerickson1 — APPROVED** [review]

**Pull request merged**